### PR TITLE
[CDAP-18648] Add a service in app fabric to scan for programs in stopping status

### PR DESCRIPTION
**Pull request by snehalakshmisha**
On the app fabric side we need a service that continuously scans the store for programs that are in Stopping status. If a program in such a state is found to be running beyond the specified `terminateTime` it should be forced to terminate.

Testing:
This log line was printed in app fabric logs for a program that was in stopping state beyond its terminateTime
```
2022-03-03 23:55:10,822 - WARN  [program run terminator:i.c.c.i.a.s.ProgramRunStatusMonitorService@118] - 
Forcing the termination of program 
program_run:default.basic_streaming_v5.-SNAPSHOT.spark.DataStreamsSparkStreaming.a6336a91-9b4c-11ec-8f0a-56f80572d6b0 
as it is past its graceful shutdown period
```

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+
+/**
+ * A service that periodically scans for program runs that are in Stopping state and
+ * force terminates them if they are running beyond the expected terminateTime.
+ */
+public class ProgramRunStatusMonitorService extends AbstractIdleService {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Extends from the `AbstractRetryableScheduledService` instead.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(ProgramRunStatus.STOPPING, 0L,
+                                                                    Long.MAX_VALUE, txBatchSize, null);
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      if (Objects.nonNull(runRecordDetail.getTerminateTs()) &&
+        runRecordDetail.getTerminateTs() <= TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())) {
+        ProgramRunId programRunId = runRecordDetail.getProgramRunId();
+        ProgramController controller = runtimeService.lookup(programRunId.getParent(),
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: The `lookup` method may return null.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(ProgramRunStatus.STOPPING, 0L,
+                                                                    Long.MAX_VALUE, txBatchSize, null);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: The endtime shouldn't be max. It just need to be up to the current time minus some grace period.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      if (Objects.nonNull(runRecordDetail.getTerminateTs()) &&
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Why throwing exception if the terminate ts is null?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        runRecordDetail.getTerminateTs() <= TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())) {
+        ProgramRunId programRunId = runRecordDetail.getProgramRunId();
+        ProgramController controller = runtimeService.lookup(programRunId.getParent(),
+                                                             RunIds.fromString(programRunId.getRun()))
+          .getController();
+        if (controller != null) {
+          LOG.warn("Forcing the termination of program {} as it is past its graceful shutdown period", programRunId);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Log with the supposed termination time too.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 1 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: As discussed offline, this is to avoid writing scheduled executors

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    long specifiedInterval = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_MILLIS);
+    if (specifiedInterval <= 0) {
+      LOG.warn("Invalid interval specified for the program terminator {}. Setting it to 5 minutes.",
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: nit: put the invalid value next to the word: Invalid interval {} specified for the program terminator.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    long specifiedInterval = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_MILLIS);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: millis seems like overkill, it's more user friendly to make this in seconds.

If this is in millis for unit tests, it would be better to better to pass in the interval into this constructor like how the txBatchSize is passed in.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * force terminates them if they are running beyond the expected terminateTime.
+ */
+public class ProgramRunStatusMonitorService extends AbstractRetryableScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunStatusMonitorService.class);
+
+  private final Store store;
+  private final ProgramRuntimeService runtimeService;
+  private final int txBatchSize;
+  private final long interval;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: nit: intervalMillis is more descriptive

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      LOG.warn("Invalid interval specified for the program terminator {}. Setting it to 5 minutes.",
+               specifiedInterval);
+      specifiedInterval = 30000L;
+    }
+    this.interval = specifiedInterval;
+  }
+
+  @Override
+  protected long runTask() {
+    LOG.debug("Starting ProgramRunStatusMonitorService");
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: this will get logged every interval which makes "Starting" misleading. The log message isn't very informative here, I think the log message that happens when a run is being terminated should be enough info.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                                 int txBatchSize) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    long specifiedInterval = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_MILLIS);
+    if (specifiedInterval <= 0) {
+      LOG.warn("Invalid interval specified for the program terminator {}. Setting it to 5 minutes.",
+               specifiedInterval);
+      specifiedInterval = 30000L;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: 30,000 is 30 seconds, not 5 minutes. Better to do `TimeUnit.MINUTES.toMillis(5)`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    return this.interval;
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state that started at most a minute before current time.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    long currentTimeInMillis = System.currentTimeMillis();
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(ProgramRunStatus.STOPPING, 0L,
+                                                                    currentTimeInMillis - 60000, txBatchSize, null);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: more readable to do `TimeUnit.MINUTES(1).toMillis()`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      Long terminateTime = runRecordDetail.getTerminateTs();
+      // If terminateTime or RuntimeInfo are null, there isn't anything that can be done about it, so we just ignore it
+      if (terminateTime != null && terminateTime <= TimeUnit.MILLISECONDS.toSeconds(currentTimeInMillis)) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: try to avoid nested ifs when possible. In cases like this, it's more readable to do:

```
for (...) {
  if (terminateTime == null || terminateTime > ...) {
    continue;
  }
  ...
  if (runtimeInfo == null) {
    continue;
  }
  controller = ...;
  if (controller != null) {
    LOG.info(...)
    controller.stop();
  }
}
```

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        RuntimeInfo runtimeInfo = runtimeService.lookup(programRunId.getParent(),
+                                                        RunIds.fromString(programRunId.getRun()));
+        if (runtimeInfo != null) {
+          ProgramController controller = runtimeInfo.getController();
+          if (controller != null) {
+            LOG.warn("Forcing the termination of program {} as it should have stopped at {} ",
+                     programRunId, terminateTime);
+            controller.stop();
+          }
+          // If controller is null it is because the program was not found during lookup.
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: nit: better to put this comment before the null check.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+      Long terminateTime = runRecordDetail.getTerminateTs();
+      // If terminateTime or RuntimeInfo are null, there isn't anything that can be done about it, so we just ignore it
+      if (terminateTime != null && terminateTime <= TimeUnit.MILLISECONDS.toSeconds(currentTimeInMillis)) {
+        ProgramRunId programRunId = runRecordDetail.getProgramRunId();
+        RuntimeInfo runtimeInfo = runtimeService.lookup(programRunId.getParent(),
+                                                        RunIds.fromString(programRunId.getRun()));
+        if (runtimeInfo != null) {
+          ProgramController controller = runtimeInfo.getController();
+          if (controller != null) {
+            LOG.warn("Forcing the termination of program {} as it should have stopped at {} ",
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: warnings usually indicate some issue with the system. In this case, appfabric is working as intended, so an info message is more fitting.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    ProgramRuntimeService testService = new AbstractProgramRuntimeService(
+      cConf, null, null, new NoOpProgramStateWriter(), null) {
+      @Nullable
+      public RuntimeInfo lookup(ProgramId programId, RunId runId) {
+        return getRuntimeInfo(programId, latch);
+      }
+    };
+    ProgramRunStatusMonitorService programRunStatusMonitorService
+      = new ProgramRunStatusMonitorService(cConf, store, testService, 5);
+    Assert.assertEquals("Latch count before stop is executed is 1", 1, latch.getCount());
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorServiceTest.java]: The message is what's displayed when the assert fails, it should be a message about what's wrong and not a message about what's expected.

In these cases I think it's ok to not give a custom message and let Assert do the default behavior, which will print both the expected and actual values.

**Review by albertshau — COMMENTED** [review]
minor comments

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 14 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Makes sense

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 12 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Thanks for the clear description :)

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize, long specifiedIntervalSecs) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    if (specifiedIntervalSecs <= 0) {
+      LOG.warn("Invalid interval {} specified for the program terminator. Setting it to 5 minutes.",
+               specifiedIntervalSecs);
+      specifiedIntervalSecs = TimeUnit.MINUTES.toMillis(5);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Should it be in seconds instead of millis?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    this.intervalMillis = TimeUnit.SECONDS.toMillis(specifiedIntervalSecs);
+  }
+
+  @Override
+  protected long runTask() {
+    terminateStoppingProgramsActiveBeyondTerminateTime();
+    return this.intervalMillis;
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Use a more concise name. E.g. terminatePrograms is good enough

Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: If you feel like explanation is needed beyond the method name, use javadoc

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    // fetch all runs that are in stopping state that started at most a minute before current time.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    long currentTimeInMillis = System.currentTimeMillis();
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: This has a risk of consuming too much memory. Better fetch in batches

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+      ProgramRunStatus.STOPPING, 0L, currentTimeInMillis - TimeUnit.MINUTES.toMillis(1), txBatchSize, null);
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      Long terminateTime = runRecordDetail.getTerminateTs();
+      // If terminateTime or RuntimeInfo are null, there isn't anything that can be done about it, so we just ignore it
+      if (terminateTime == null || terminateTime > TimeUnit.MILLISECONDS.toSeconds(currentTimeInMillis)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: The current time in seconds is an invariant , should just compute once outside of the loop.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 20]
```diff
@@ -562,6 +562,66 @@
...
+    <name>app.program.terminator.interval.secs</name>
+    <value>300</value>
+    <description>
+      Interval in seconds of how often the program terminator thread will run;
+      this value should be greater than 0
+    </description>
+  </property>
+
+  <property>
+    <name>program.run.status.monitor.service.retry.policy.base.delay.ms</name>
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml]: This is a very long name. I believe we already have a retry policy for runtime monitor, can we just use that?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 16 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Oops. My bad

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 21]
```diff
@@ -1271,46 +1271,53 @@ public void getStatuses(FullHttpRequest request, HttpResponder responder,
   @Path("/stop")
   @AuditPolicy({AuditDetail.REQUEST_BODY, AuditDetail.RESPONSE_BODY})
   public void stopPrograms(FullHttpRequest request, HttpResponder responder,
-                           @PathParam("namespace-id") String namespaceId) throws Exception {
-
+                           @PathParam("namespace-id") String namespaceId,
+                           @QueryParam("graceful") String gracefulShutdownSecs) throws Exception {
+    LOG.info("Stopping all programs in namespace {}", namespaceId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: was this added intentionally? Seems unrelated

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java [hunk 22]
```diff
@@ -873,6 +873,112 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
...
+   * @param programId
+   * @param gracefulShutdownSecs
+   * @param runId
+   * @return
+   * @throws BadRequestException
+   * @throws NotFoundException
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public List<ProgramRunId> publishStoppingToTms(ProgramId programId, int gracefulShutdownSecs,
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java]: same here, accidentally included? I'll be refactoring this as part of the program stop change, you don't need to handle it.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 21 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: I made changes to this file and the ProgramLifecycleService files to validate my changes. Will remove this commit once I am done testing

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java (hunk 22 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java]: Made these changes for testing only

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 19 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: currentTime was initialized outside loop but the conversion was being done here. Moved the conversion as well to outside the loop

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 18 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Was already fetching in batches. Using a filter now to filter more runs based on terminateTime

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+        if (runtimeInfo != null && runtimeInfo.getController() != null) {
+          LOG.info("Forcing the termination of program {} as it should have stopped at {} ",
+                   programRunId, record.getTerminateTs());
+          runtimeInfo.getController().stop();
+        }
+      }
+    }
+  }
+
+  private Predicate<RunRecordDetail> createFilter(long currentTimeInSecs) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: this filter will keep returning the same thing until the RuntimeInfo is removed from runtimeService. Since the kill is not immediate, what will happen is the same set of run ids will be returned for each scan and get repeatedly killed.

The way it is done in the RunRecordCorrector is that there is a Set of program run ids, such that each store.getRuns() call will not return overlapping runs.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+    return record -> {
+      Long terminateTime = record.getTerminateTs();
+      if (terminateTime == null || terminateTime > currentTimeInSecs) {
+        return false;
+      }
+      ProgramRunId programRunId = record.getProgramRunId();
+      RuntimeInfo runtimeInfo = runtimeService.lookup(programRunId.getParent(),
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: this is checked in the terminatePrograms() loop, there is no need to do it here.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java (hunk 23 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java]: Ahh I see

**Review by albertshau — APPROVED** [review]
lgtm, though there are conflicts that need to be fixed
